src/v/cloud_storage/materialized_resources.cc

```diff
@@ -115,84 +105,159 @@ ss::future<> materialized_resources::start() {
+    static size_t sz = remote_partition::reader_mem_use_estimate();
+    return static_cast<ssize_t>(sz);
+}
+
+inline ssize_t projected_remote_segment_memory_usage() {
+    // This is an estimate. When the reader is created it's size should be
+    // checked and if it's lareger units should be acquired. If it's smaller
+    // some units should be released.
+    static constexpr size_t segment_index_projected_size = 0x1000;
+    return sizeof(remote_segment) + segment_index_projected_size;
+}
+
 materialized_manifest_cache&
 materialized_resources::get_materialized_manifest_cache() {
     return *_manifest_cache;
 }
 
 size_t materialized_resources::max_segment_readers() const {
-    return static_cast<size_t>(_max_segment_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
+    // TODO: implement override logic
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_segment_reader_memory_usage();
 }
 
 size_t materialized_resources::max_partition_readers() const {
-    return static_cast<size_t>(_max_partition_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
+    // TODO: fixme
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_partition_reader_memory_usage();
 }
 
 size_t materialized_resources::max_segments() const {
-    return static_cast<size_t>(_max_segments_per_shard().value_or(
-      _max_partitions_per_shard() * default_segment_factor));
+    // TODO: fixme
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_segment_memory_usage();
 }
 
 size_t materialized_resources::current_segment_readers() const {
-    return _segment_reader_units.outstanding();
+    size_t n = 0;
+    for (const auto& s : _materialized) {
+        n += s.readers.size();
+    }
+    return n;
 }
 
 size_t materialized_resources::current_partition_readers() const {
-    return _partition_reader_units.outstanding();
+    // TODO: fixme
+    return 0;
 }
 
 size_t materialized_resources::current_segments() const {
-    return _segment_units.outstanding();
+    return _materialized.size();
 }
 
 void materialized_resources::register_segment(materialized_segment_state& s) {
+    // We can only estimate the actual memory usage after the object is
+    // created. We're using projected memory useage to actually acquire the
+    // units. To fix this we need to check the acquired units and adjust
+    // them. If the units overshoot we need to return some. Otherwise we
+    // need to adopt extra units. When this happens the semaphore counter
+    // may go negative. This means that we used more memory than needed.
+    // In this case we need to mark the segment using 'overcommit' flag
+    // so next eviction will remove it.
+    auto estimate = s.segment->estimate_memory_use();
+    auto units = s._units.count();
+    if (units > estimate) {
```

Comment:
Is it possible that the two values will ever be equal? Do we need to do nothing in that case?

Reply:
yes, it will try to consume 0 units and `checkpoint_hint` will be set to false IIUC